torch/csrc/jit/passes/concat_opt.cpp

```diff
@@ -0,0 +1,487 @@
+      return;
+    }
+    if (!allShapesAreKnown(node)) {
+      // Can't expand when shapes are not known for the `cat` op.
+      return;
+    }
+    for (auto cat_inp : node->input(0)->node()->inputs()) {
+      if (!shapeIsKnown(cat_inp)) {
+        // Can't expand when shapes of the inputs to `cat` are not known.
+        return;
+      }
+    }
+    auto cat_dim = node->input(1);
+    if (!cat_dim->node()->hasAttribute(attr::value)) {
+      // Can't expand when cat dimension is not a constant.
+      return;
+    }
+    int cat_dim_value = cat_dim->node()->i(attr::value);
+
+    // Set the insertion point to the curent `cat` node.
+    WithInsertPoint guard(node);
+    auto none = graph_->insertConstant(IValue());
+    auto one = graph_->insertConstant(1);
+    auto empty_symbol = aten::empty;
+    auto slice_symbol = aten::slice;
+    auto copy_symbol = aten::copy_;
+
+    // Insert the constants needed for the `cat` output buffer size.
+    auto tensortype = node->output()->type()->expect<TensorType>();
+    TORCH_INTERNAL_ASSERT(tensortype);
+    auto tensortype_sizes = tensortype->sizes();
+    std::vector<Value*> cat_out_size;
+    for (size_t i = 0; i < tensortype_sizes.size(); ++i) {
+      cat_out_size.push_back(graph_->insertConstant(tensortype_sizes[i]));
+    }
+
+    // Create a list of int for `cat` output buffer size.
+    auto cat_out_size_list = graph_->createList(IntType::get(), cat_out_size);
+    cat_out_size_list->insertBefore(node);
+
+    // Create an empty buffer to be used as `cat` output buffer.
+    auto cat_out_empty = graph_->create(
+        empty_symbol,
+        {cat_out_size_list->output(), none, none, none, none, none});
+    cat_out_empty->insertBefore(node);
+
+    // For every input to this `cat` node:
+    //   * Create a slice of `cat` output buffer.
+    auto cat_out_value = cat_out_empty->output();
+    auto cat_inp_list = node->input(0)->node();
+    int start_idx = 0;
+    auto start = graph_->insertConstant(start_idx);
+    for (auto cat_inp : cat_inp_list->inputs()) {
+      // Create a slice of the cat output buffer that correspond to
+      // this input size and position in the output.
+      auto cat_inp_tensor_type =
+          dynamic_cast<TensorType*>(cat_inp->type().get());
+      TORCH_INTERNAL_ASSERT(cat_inp_tensor_type);
+      TORCH_INTERNAL_ASSERT(cat_inp_tensor_type->dim());
+      auto cat_inp_tensortype_sizes = cat_inp_tensor_type->sizes();
+      int end_idx = start_idx + *cat_inp_tensortype_sizes[cat_dim_value];
+      auto end = graph_->insertConstant(end_idx);
+
+      auto slice = graph_->create(
+          slice_symbol, {cat_out_value, cat_dim, start, end, one});
+      GRAPH_UPDATE("Inserting\n", *slice, "before\n", *node);
+      slice->insertBefore(node);
+      slices_added_.push_back(slice);
+
+      // Insert a copy from this input to the output slice.
+      auto copy = graph_->create(copy_symbol, {slice->output(), cat_inp});
```

Comment:
nit: I think it would be more readable if `copy_symbol` was inline. as is, i had to look back to double check that this was `copy_` and not `copy`. (the second one doesnt exist, but i didnt know that off hand reading the code)

Reply:
Updated.